extensions/aerosol_optics/mo_aerosol_optics.F90

```diff
@@ -0,0 +1,633 @@
+                                               ! 3 = merra_aero_sulf    (sulfate)
+                                               ! 4 = merra_aero_bcar_rh (black carbon, hydrophilic)
+                                               ! 5 = merra_aero_bcar    (black carbon, hydrophobic)
+                                               ! 6 = merra_aero_ocar_rh (organic carbon, hydrophilic)
+                                               ! 7 = merra_aero_ocar    (organic carbon, hydrophobic)
+    real(wp), intent(in  ) :: aero_size(:,:)   ! aerosol size for dust and sea-salt
+                                               ! Dimensions: (ncol,nlay)
+    real(wp), intent(in  ) :: aero_mass(:,:)   ! aerosol mass column (g/m2)
+                                               ! Dimensions: (ncol,nlay)
+    real(wp), intent(in  ) :: relhum(:,:)      ! relative humidity (fraction, 0-1)
+                                               ! Dimensions: (ncol,nlay)
+
+    class(ty_optical_props_arry), &
+              intent(inout) :: optical_props
+                                               ! Dimensions: (ncol,nlay,nbnd)
+
+    character(len=128)      :: error_msg
+    ! ------- Local -------
+    logical(wl), dimension(size(aero_type,1), size(aero_type,2)) :: aeromsk
+    real(wp),    dimension(size(aero_type,1), size(aero_type,2), size(this%aero_dust_ext,2)) :: &
+                 atau, ataussa, ataussag
+    integer  :: ncol, nlay, nbnd, nbin, nrh
+    integer  :: icol, ilay, ibnd, ibin
+    ! scalars for total tau, tau*ssa
+    real(wp) :: tau, taussa
+
+    ! ----------------------------------------
+    !
+    ! Error checking
+    !
+    ! ----------------------------------------
+
+    error_msg = ''
+    if(.not.(allocated(this%aero_dust_ext))) then
+      error_msg = 'aerosol optics: no data has been initialized'
+      return
+    end if
+
+    ncol = size(aero_type,1)
+    nlay = size(aero_type,2)
+    nbin = size(this%aero_bin_lo,1)
+    nrh = size(this%aero_rh,1)
+    nbnd = size(this%aero_dust_ext,2)
+
+    !
+    ! Array sizes
+    !
+    if (check_extents) then
+      if(size(aero_type, 1) /= ncol .or. size(aero_type, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_type has wrong extents"
+      if(size(aero_size, 1) /= ncol .or. size(aero_size, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_size has wrong extents"
+      if(size(aero_mass, 1) /= ncol .or. size(aero_mass, 2) /= nlay) &
+        error_msg = "aerosol optics: aero_mass has wrong extents"
+      if(size(relhum,  1) /= ncol .or. size(relhum,  2) /= nlay) &
+        error_msg = "aerosol optics: relhumn has wrong extents"
+      if(optical_props%get_ncol() /= ncol .or. optical_props%get_nlay() /= nlay) &
+        error_msg = "aerosol optics: optical_props have wrong extents"
+      if(error_msg /= "") return
+    end if
+
+    !
+    ! Spectral consistency
+    !
+    if(check_values) then
```

Comment:
@mjiacono It seems that if users provide an integer value of aerosol_type that isn't in their list those entries get ignored. Is that the behavior you want? You could also add a check here to be sure the aerosol_type varies from 0 to the max value of aerosol type.

Reply:
@RobertPiincus Check on aerosol type added. Error returned if out of bounds.